### PR TITLE
Added support for importing from data files from s3

### PR DESCRIPTION
**Pull request by rahulb-yb**
This PR introduces:
- A new interface `datastore.go`, which takes into consideration the location of the stored data files. A lot of the operations on locally stored files are now abstracted out through this interface to support the respective API calls needed to use S3 objects.
- S3 utility packages, which provide helper functions dedicated to making use of the API calls and keeps the rest of the code reader-friendly
- ~~S3 Download Streamer, which allows us to "stream" data from S3 objects into a readCloser interface by making the API calls believe that we are writing to a "file", which is just a pipe in our memory!~~ Introduced the usage of the go-cloud CDK, which removes the need for our own data streamer. No data from objects downloaded onto your local machine!
- Some code refactoring made in favour of the newly introduced datastore interface

To be added:
- Bucket "directory" structure support
- Custom profiles to access AWS with (we currently take the `default` profile in the aws config file)

### CODE REVIEW

File: yb-voyager/src/datastore/s3Datastore.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	return headObject.ContentLength, nil
+}
+
+// filepath.Join converts URL (s3://...) to directory-like string (s3:/...).
+func (ds *S3Datastore) Join(elem ...string) string {
+	finalPath := ""
+	for _, fragment := range elem {
+		finalPath += fragment + "/"
+	}
+	return finalPath[:len(finalPath)-1]
```
Comment (amit-yb) [inline yb-voyager/src/datastore/s3Datastore.go:63]: `strings.Join()`?

File: yb-voyager/src/datastore/s3Datastore.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (ds *S3Datastore) Open(resourceName string) (io.ReadCloser, error) {
+	// resourceName is hidden underneath a symlink for s3 objects.
+	s3Resource, err := os.Readlink(resourceName)
+	if err != nil {
+		utils.ErrExit("unable to resolve symlink %v to s3 resource: %v", resourceName, err)
+	}
+	reader, err := s3.DownloadS3Object(s3Resource)
```
Comment (amit-yb) [inline yb-voyager/src/datastore/s3Datastore.go:72]: How about renaming `DownloadS3Object()` to something like `OpenObjectForReading()`?

File: yb-voyager/src/utils/s3/s3utils.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	"gocloud.dev/blob/s3blob"
+)
+
+var client *s3.Client
+
+func createClientIfNotExists() {
+	if client != nil {
+		return
+	}
+	cfg, err := config.LoadDefaultConfig(context.TODO())
```
Comment (amit-yb) [inline yb-voyager/src/utils/s3/s3utils.go:38]: `context.Background()`

File: yb-voyager/src/utils/s3/s3utils.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	if err != nil {
+		utils.ErrExit("load s3 config: %v", err)
+	}
+	client = s3.NewFromConfig(cfg)
+}
+
+func VerifyS3FromDataDir(datadir string) error {
```
Comment (amit-yb) [inline yb-voyager/src/utils/s3/s3utils.go:45]: The name of the function is not matching its implementation.

File: yb-voyager/src/utils/s3/s3utils.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+		return err
+	}
+	bucket := u.Host
+	if bucket == "" {
+		return fmt.Errorf("missing bucket in s3 url")
+	}
+	return nil
+}
+
+func S3FromUrl(resourceName string) (string, string, error) {
```
Comment (amit-yb) [inline yb-voyager/src/utils/s3/s3utils.go:57]: Rename to `SplitObjectPath()`.
Rename `resourceName` to `objectPath`.

File: yb-voyager/src/utils/s3/s3utils.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	u, err := url.Parse(resourceName)
+	if err != nil {
+		return "", "", err
+	}
+	bucket := u.Host
+	key := u.Path[1:] //remove initial "/", unable to find object with it
+	if bucket == "" {
+		return "", "", fmt.Errorf("missing bucket in s3 url")
```
Comment (amit-yb) [inline yb-voyager/src/utils/s3/s3utils.go:65]: Include the `objectPath` in the error message. Same for the error on line 68.

File: yb-voyager/cmd/importDataFileCommand.go [hunk 7]
```diff
@@ -71,11 +74,11 @@ func prepareForImportDataCmd() {
 func getFileSizeInfo() map[string]int64 {
 	tableFileSize := make(map[string]int64)
 	for table, filePath := range tableNameVsFilePath {
-		fileInfo, err := os.Stat(filePath)
+		fileBytes, err := ds.FileSize(filePath)
```
Comment (amit-yb) [inline yb-voyager/cmd/importDataFileCommand.go:77]: fileBytes ==> fileSize ??

File: yb-voyager/cmd/importDataFileCommand.go [hunk 8]
```diff
@@ -149,17 +156,19 @@ func setImportTableListFlag() {
 }
 
 func parseFileTableMapping() {
+	ds = datastore.NewDataStore(dataDir)
```
Comment (amit-yb) [inline yb-voyager/cmd/importDataFileCommand.go:159]: This doesn't look like a right place for this.
Initialise `ds` as a first statement in the `Run()` method of both `import data` and `import data file`. So, the rest of the code paths can always assume that `ds` is already initialised.

**Review by amit-yb — COMMENTED** [review]
Just a few minor comments.

File: yb-voyager/src/utils/s3/s3utils.go (hunk 4 above)
Comment (rahulb-yb) [reply yb-voyager/src/utils/s3/s3utils.go:45]: We are simply verifying that the given s3 URL is not a malformed one, is there something else I should be naming this function?

File: yb-voyager/src/utils/s3/s3utils.go [hunk 9]
```diff
@@ -49,23 +49,23 @@ func VerifyS3FromDataDir(datadir string) error {
 	}
 	bucket := u.Host
 	if bucket == "" {
-		return fmt.Errorf("missing bucket in s3 url")
+		return fmt.Errorf("missing bucket in s3 url %v", datadir)
```
Comment (amit-yb) [inline yb-voyager/src/utils/s3/s3utils.go:52]: Always use `%w` for wrapping errors.

File: yb-voyager/src/datastore/s3Datastore.go [hunk 10]
```diff
@@ -55,20 +55,18 @@ func (ds *S3Datastore) FileSize(filePath string) (int64, error) {
...
 
-func (ds *S3Datastore) Open(resourceName string) (io.ReadCloser, error) {
+func (ds *S3DataStore) Open(resourceName string) (io.ReadCloser, error) {
 	// resourceName is hidden underneath a symlink for s3 objects.
 	s3Resource, err := os.Readlink(resourceName)
 	if err != nil {
 		utils.ErrExit("unable to resolve symlink %v to s3 resource: %v", resourceName, err)
 	}
-	reader, err := s3.DownloadS3Object(s3Resource)
+	reader, err := s3.OpenReadableObject(s3Resource)
```
Comment (amit-yb) [inline yb-voyager/src/datastore/s3Datastore.go:70]: Change `s3Resource` to `objectPath`.

Change `OpenReadableObject()` to something like `NewObjectReader(objectPath)` or `OpenObjectForReading(objectPath)`.

**Review by amit-yb — APPROVED** [review]
LGTM. Just a few minor comments.

File: yb-voyager/src/utils/s3/s3utils.go (hunk 4 above)
Comment (amit-yb) [reply yb-voyager/src/utils/s3/s3utils.go:45]: Your above response itself gives the name: `validateObjectURL(...)`

**Pull request merged**